source/gameengine/Ketsji/KX_CubeMapManager.cpp

```diff
@@ -0,0 +1,157 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ * Contributor(s): Ulysse Martin, Tristan Porteries.
+ *
+ * ***** END GPL LICENSE BLOCK *****
+ */
+
+/** \file gameengine/Ketsji/KX_CubeMapManager.cpp
+ *  \ingroup ketsji
+ */
+
+#include "KX_CubeMapManager.h"
+#include "KX_Camera.h"
+#include "KX_Scene.h"
+#include "KX_Globals.h"
+#include "KX_CubeMap.h"
+
+#include "EXP_ListValue.h"
+
+#include "RAS_IRasterizer.h"
+
+KX_CubeMapManager::KX_CubeMapManager(KX_Scene *scene)
+	:m_scene(scene)
+{
+	const RAS_CameraData& camdata = RAS_CameraData();
+	m_camera = new KX_Camera(m_scene, KX_Scene::m_callbacks, camdata);
+}
+
+KX_CubeMapManager::~KX_CubeMapManager()
+{
+	for (std::vector<KX_CubeMap *>::iterator it = m_cubeMaps.begin(), end = m_cubeMaps.end(); it != end; ++it) {
+		delete *it;
+	}
+
+	delete m_camera;
+}
+
+void KX_CubeMapManager::AddCubeMap(KX_CubeMap *cubeMap)
+{
+	m_cubeMaps.push_back(cubeMap);
+}
+
+void KX_CubeMapManager::InvalidateCubeMapViewpoint(KX_GameObject *gameobj)
+{
+	for (std::vector<KX_CubeMap *>::iterator it = m_cubeMaps.begin(), end = m_cubeMaps.end(); it != end; ++it) {
+		KX_CubeMap *cubeMap = *it;
+		if (cubeMap->GetViewpointObject() == gameobj) {
+			cubeMap->SetViewpointObject(NULL);
+		}
+	}
+}
+
+void KX_CubeMapManager::RenderCubeMap(RAS_IRasterizer *rasty, KX_CubeMap *cubeMap)
+{
+	KX_GameObject *viewpoint = cubeMap->GetViewpointObject();
+
+	// Doesn't need (or can) update.
+	if (!cubeMap->NeedUpdate() || !cubeMap->GetEnabled() || !viewpoint) {
+		return;
+	}
+
+	const MT_Vector3& position = viewpoint->NodeGetWorldPosition();
+
+	/* We hide the viewpoint object in the case backface culling is disabled -> we can't see through
+	 * the object faces if the camera is inside the gameobject.
+	 */
+	viewpoint->SetVisible(false, true);
+
+	// For Culling we need first to set the camera position at the object position.
+	m_camera->NodeSetWorldPosition(position);
+
+	/* When we update clipstart or clipend values,
+	 * or if the projection matrix is not computed yet,
+	 * we have to compute projection matrix.
+	 */
+	if (cubeMap->GetInvalidProjectionMatrix()) {
+		const float clipstart = cubeMap->GetClipStart();
+		const float clipend = cubeMap->GetClipEnd();
+		const MT_Matrix4x4& proj = rasty->GetFrustumMatrix(-clipstart, clipstart, -clipstart, clipstart, clipstart, clipend, 1.0f, true);
+		cubeMap->SetProjectionMatrix(proj);
+		cubeMap->SetInvalidProjectionMatrix(false);
+	}
+
+	// Else we use the projection matrix stored in the cube map.
+	rasty->SetProjectionMatrix(cubeMap->GetProjectionMatrix());
+	m_camera->SetProjectionMatrix(cubeMap->GetProjectionMatrix());
+
+	cubeMap->BeginRender();
```

Comment:
There is not a problem here? Do we really need to CheckValidTexture each frame? When I remove that I have a performance increase.